src/Bundle.ts

```diff
@@ -12,68 +12,349 @@ import transformBundle from './utils/transformBundle';
 	}
 
-	includeMarked (treeshake: boolean) {
-		if (treeshake) {
-			let addedNewNodes;
-			do {
-				addedNewNodes = false;
-				this.orderedModules.forEach(module => {
-					if (module.includeInBundle()) {
-						addedNewNodes = true;
-					}
-				});
-			} while (addedNewNodes);
+	// ensure that the module exports or reexports the given variable
+	// we don't replace reexports with the direct reexport from the final module
+	// as this might result in exposing an internal module which taints an entryModule bundle
+	ensureExport (module: Module | ExternalModule, variable: Variable): string {
+		let safeExportName = this.exportedVariables.get(variable);
+		if (safeExportName) {
+			return safeExportName;
+		}
+
+		let i = 0;
+		if (variable.exportName) {
+			safeExportName = variable.exportName;
 		} else {
-			// Necessary to properly replace namespace imports
-			this.orderedModules.forEach(module => module.includeAllInBundle());
+			safeExportName = variable.exportName = variable.name;
+		}
+
+		let curExport = this.exports[safeExportName];
+		while (curExport) {
+			safeExportName = (variable.exportName || variable.name) + '$' + ++i;
+			curExport = this.exports[safeExportName];
+		}
+
+		curExport = this.exports[safeExportName] = { module, name: undefined, variable };
+		this.exportedVariables.set(variable, safeExportName);
+
+		// if we've just exposed an export of a non-entry-point,
+		// then note we are no longer an entry point bundle
+		// we will then need an entry point facade if this is an entry point module
+		if (this.entryModuleFacade && module.bundle === this && !module.isEntryPoint) {
+			this.entryModuleFacade = false;
+		}
+
+		// if we are reexporting a module in another bundle
+		// then we also have to ensure it is an export there too
+		// and note the name it comes from
+		if (module.bundle !== this && !module.isExternal) {
+			curExport.name = (<Module>module).bundle.ensureExport(module, variable);
 		}
+		else {
+			curExport.name = safeExportName;
+		}
+
+		return safeExportName;
 	}
 
-	setOutputFacade (entryModule: Module) {
-		this.entryModule = entryModule;
+	generateEntryExports (entryModule: Module) {
+		entryModule.getAllExports().forEach(exportName => {
+			const traced = this.traceExport(entryModule, exportName);
+			const variable = traced.module.traceExport(traced.name);
+			this.exports[exportName] = { module: traced.module, name: traced.name, variable };
+			// if we exposed an export in another module ensure it is exported there
+			if (traced.module.bundle !== this && !traced.module.isExternal) {
+				(<Module>traced.module).bundle.ensureExport(traced.module, variable);
+			}
+			this.exportedVariables.set(variable, exportName);
+		});
 	}
 
-	processExternals () {
-		// prune unused external imports
-		this.externalModules = this.graph.externalModules.filter(module => {
-			return module.used || !this.graph.isPureExternalModule(module.id);
+	generateDependencies (entryFacade?: Module) {
```

Comment:
Isn't it more like `collectDependencies`?

Reply:
Sure, I was struggling to name these functions actually.